### PR TITLE
Adaptive Eviction in Neuron-level model's memory - Arch Improvement Idea - Open to feedback

### PR DESCRIPTION
**Pull request by Mystic-Slice**
Hi!
I was reading the CTM paper and found it interesting! I was playing around with the code to explore the new architecture more when I got this idea. The idea, admittedly is very rough. I am running some experiments to explore the idea but I dont yet conclusive know if it is a useful addition to the architecture.
I will share my thoughts in this PR. I am happy to receive comments / feedback on this.

### CODE REVIEW

**Comment by Mystic-Slice** [comment]
CTM preserves a brief history of post-activations for each of the neuron-level model.
Currently, at each tick, the oldest memory is taken out to make space for the recent one.
Would there be an advantage to allowing the model to choose which memory to evict?

I implented an 'evictor' network that decides which memory to kick out. This currently uses only the memory sequence as the input.

**My thoughts:**
1. Adaptive Eviction would allow preservation of more important memories for the next tick. The model can kick out what it thinks is the least important memory.
2. This makes the most sense to me in tasks requiring sequential thinking. The model could pursue a single branch in its tree of possible solutions. And here, when it decides the current solution is not the suitable, it can evict these memories while preserving the top level memory still to pursue another branch of the solution.
3. It can also possibly allow neuron-level 'summaries' to be preserved in the memory when going on large number of iterations.

**Ideas to improve it further:**
1. Include the new memory/state as an input for this classifier. The model can make a more informed decision when it knows what is replace the memory being kicked out.

**Comment by Mystic-Slice** [comment]
This works well for sorting task.
I ran the training for `N=30, training_iterations=50000` with and without adaptive eviction with all other settings set to the default ones provided by the authors.

>Note: I believe the model doesn't converge within 50000 steps but that's all I could train for given my compute constraints. I acknowledge that this is a major weakness of this experiment and the actual results after training it to convergence might be very different.

I observed a `>10%` test accuracy improvement (32.3% [max: 38.5%] -> 45.8% [max: 51%]) for sorting the entire list correctly with adaptive eviction while it contributes 10% increase in number of parameters (897197 -> 999766 [+102569]). The parameter count of adaptive eviction increases with the memory length and the dim of neuron-level models' output.

**Training Acc curve:**
image

**Test Acc curve:**
image

**Training Loss curve:**
image

**Test Loss curve:**
image

**Test Loss curve (zoomed in to the last 5000 iterations):**
image

I also tried for `N=10, 15` but the improvement in accuracy and loss was not meaningful. I assume it was because the problem is just too easy. Both versions of CTM got upto 80% full list sorting accuracy within 10000 training iterations.

**Comment by Mystic-Slice** [comment]
Although it worked well for sorting task, it didn't show a difference in Image Classification on Cifar10 dataset.

I trained the models for `training_iterations=50000` with and without adaptive eviction. Although the loss was lower in the model with adaptive eviction, the difference is not significant. The test accuracy of the base model was as good as if not better than the one with adaptive eviction.

**Possible reasons:**
1. Adaptive Eviction is just not better
2. Adaptive Eviction is not good / needed for Image classification 
3. The training process is not long enough for us to say anything conclusive. 
         - The recommended number of training iterations is 600000 for Cifar10.
         - My experiment is not even close to the recommended training length.
         - Even in the sorting experiment (which was closer to 1/2 of the recommended training length), the first 30-40% of the training iterations, adaptive eviction model didnt show an advantage over the base model. 
         - Maybe the model just needs more time to learn how to use adaptive eviction.

**Comment by Mystic-Slice** [comment]
A friend pointed out that the code contains an undifferentiable op. A very silly mistake on my part.
The eviction network is not actually training. The results are just due to random chance.

When I recorded the evicted indices, I see that the model just evicts the same memory every tick for every input.
